source/common/network/dns_impl.cc

```diff
@@ -81,30 +81,35 @@ void DnsResolverImpl::PendingResolution::onAresHostCallback(int status, int time
 
   std::list<Address::InstanceConstSharedPtr> address_list;
   if (status == ARES_SUCCESS) {
-    if (hostent->h_addrtype == AF_INET) {
-      for (int i = 0; hostent->h_addr_list[i] != nullptr; ++i) {
-        ASSERT(hostent->h_length == sizeof(in_addr));
-        sockaddr_in address;
-        memset(&address, 0, sizeof(address));
-        address.sin_family = AF_INET;
-        address.sin_port = 0;
-        address.sin_addr = *reinterpret_cast<in_addr*>(hostent->h_addr_list[i]);
-        address_list.emplace_back(new Address::Ipv4Instance(&address));
-      }
-    } else if (hostent->h_addrtype == AF_INET6) {
-      for (int i = 0; hostent->h_addr_list[i] != nullptr; ++i) {
-        ASSERT(hostent->h_length == sizeof(in6_addr));
-        sockaddr_in6 address;
-        memset(&address, 0, sizeof(address));
-        address.sin6_family = AF_INET6;
-        address.sin6_port = 0;
-        address.sin6_addr = *reinterpret_cast<in6_addr*>(hostent->h_addr_list[i]);
-        address_list.emplace_back(new Address::Ipv6Instance(address));
+    if (addrinfo != nullptr && addrinfo->nodes != nullptr) {
+      if (addrinfo->nodes->ai_family == AF_INET) {
+        for (const ares_addrinfo_node* ai = addrinfo->nodes; ai != nullptr; ai = ai->ai_next) {
+          sockaddr_in address;
+          memset(&address, 0, sizeof(address));
+          address.sin_family = AF_INET;
+          address.sin_port = 0;
+          address.sin_addr = reinterpret_cast<sockaddr_in*>(ai->ai_addr)->sin_addr;
+          address_list.emplace_back(new Address::Ipv4Instance(&address));
+        }
+      } else if (addrinfo->nodes->ai_family == AF_INET6) {
+        for (const ares_addrinfo_node* ai = addrinfo->nodes; ai != nullptr; ai = ai->ai_next) {
+          sockaddr_in6 address;
+          memset(&address, 0, sizeof(address));
+          address.sin6_family = AF_INET6;
+          address.sin6_port = 0;
+          address.sin6_addr = reinterpret_cast<sockaddr_in6*>(ai->ai_addr)->sin6_addr;
+          address_list.emplace_back(new Address::Ipv6Instance(address));
+        }
       }
     }
+
     if (!address_list.empty()) {
       completed_ = true;
     }
+
+    if (addrinfo) {
```

Comment:
Nit: should this just be `ASSERT(addrinfo != nullptr);`?

Reply:
Not sure if the lib guarantees that `addrinfo` is not `nullptr` when status is success. I would like to keep `if`.

Reply:
I think it does, see https://github.com/c-ares/c-ares/blob/master/ares_getaddrinfo.3#L126.

Reply:
I will replace with the `ASSERT`.